### ISSUE TITLE
Transpiling Shor circuits bottlenecked by operator class operations

### ISSUE THREAD

**Issue by nonhermitian** [Qiskit/qiskit-terra#5996]
### What is the expected enhancement?

Transpiling even a "small" Shor circuit takes a long time:

```python

from qiskit.aqua.algorithms import Shor
N = 5
shor = Shor(N)
sh = shor.construct_circuit()
trans_qc = transpile(sh, backend)
```
Small of course is quite large in this case, 14Q to depth ~7k, but it also highlights that a large amount of time is spent doing simple things like init Operators.  Indeed, init, append, set_dims, and compose Operator methods take 15 out of 19 seconds of the transpilation time

```
         36519184 function calls (35886556 primitive calls) in 19.824 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   167578    0.816    0.000    1.475    0.000 dagcircuit.py:249(_add_wire)
   285076    0.739    0.000    0.739    0.000 {method 'reduce' of 'numpy.ufunc' objects}
     9097    0.605    0.000    1.518    0.000 dagcircuit.py:882(substitute_node_with_dag)
    47134    0.474    0.000    1.012    0.000 dagcircuit.py:352(apply_operation_back)
   300034    0.401    0.000    1.237    0.000 copy.py:66(copy)
   247688    0.392    0.000    1.283    0.000 fromnumeric.py:70(_wrapreduction)
186645/170227    0.391    0.000    1.092    0.000 copy.py:269(_reconstruct)
636456/367752    0.347    0.000    3.012    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
  3405116    0.343    0.000    0.382    0.000 {built-in method builtins.isinstance}
   428374    0.337    0.000    0.552    0.000 dagnode.py:32(__init__)
   268253    0.325    0.000    0.325    0.000 {built-in method numpy.array}
   167578    0.285    0.000    0.285    0.000 {method 'add_nodes_from' of 'retworkx.PyDiGraph' objects}
     7760    0.283    0.000    0.283    0.000 {qiskit.transpiler.passes.routing.cython.stochastic_swap.swap_trial.swap_trial}
     5197    0.277    0.000    1.022    0.000 dagcircuit.py:593(compose)
   114748    0.252    0.000    0.654    0.000 base_operator.py:408(_automatic_dims)
   444090    0.241    0.000    0.330    0.000 register.py:105(__getitem__)
    15928    0.227    0.000    0.286    0.000 {lexicographical_topological_sort}
   728230    0.212    0.000    0.215    0.000 {built-in method builtins.getattr}
57374/4768    0.204    0.000    6.229    0.001 operator.py:50(__init__)
28687/4768    0.200    0.000    4.774    0.001 operator.py:510(_append_instruction)
    12468    0.197    0.000    1.945    0.000 dagcircuit.py:227(add_qreg)
   116377    0.188    0.000    1.014    0.000 quantumcircuit.py:843(_append)
   184385    0.187    0.000    0.187    0.000 {built-in method __new__ of type object at 0x5610d20e86e0}
    23919    0.184    0.000    3.100    0.000 operator.py:219(compose)
  2570295    0.183    0.000    0.183    0.000 bit.py:73(__hash__)
    57374    0.180    0.000    1.119    0.000 base_operator.py:422(_set_dims)
    40241    0.176    0.000    0.382    0.000 rz.py:104(to_matrix)
        9    0.175    0.019    0.468    0.052 dagcircuit.py:1425(_collect_1q_runs)
173130/11789    0.173    0.000    0.635    0.000 copy.py:132(deepcopy)
     9536    0.170    0.000    1.948    0.000 circuit_to_instruction.py:21(circuit_to_instruction)
        9    0.165    0.018   10.262    1.140 optimize_1q_decomposition.py:59(run)
```

**Comment by mtreinish** [comment]
Which version of terra are you using and which backend? If you're using the released version (0.16.x) and an ibmq backend (or any backend where the basis is not u1, u2, u3, cx) the 1q optimization pass is inefficient and slow because of it's use of operators and other bottlenecks in the 1q synthesis module. The version of that pass on master is significantly faster (see #5682, #5909, and #5928) and should eliminate the bottleneck.

If you're using master I'll dig into it some more because there are probably other places we've missed.

**Comment by mtreinish** [comment]
Ok, then I'll close this as fixed for the upcoming 0.17.0 release. Feel free to reopen it if needed or there is more to do here.
